keps/sig-api-machinery/20190324-declatative-field-manager.md

```diff
@@ -0,0 +1,138 @@
+this KEP aims to provide a declarative way of setting the current fieldManager through a field in the object itself.
+
+## Motivation
+
+It is a common practice to use Kubernetes manifests as stored configuration of an application. Either to just introduce version control to the cluster's state, or to automate the process of updating objects.
+
+With the addition of Server Side Apply it could become necessary to change the way those manifests get applied, as the most benefit from this new feature requires every actor to define it's fieldManager. Doing this, currently is only possible through api options, the user-agent or as a kubectl option.
+
+To keep the way of interacting with manifests as easy as possible, this KEP suggests to allow defining the fieldManager as part of the manifest itself. Therefor no further action is required when applying a set of manifests and users can easily benefit from Server Side Apply without having to change their apply workflow.
+
+This would also assist existing solutions already interacting with manifest files, as the fieldManager information gets provided in a way they already know to handle which might improve compatibility.
+
+### Goals
+
+- Allow users to set their fieldManager information declaratively in their manifests
+
+### Non-Goals
+
+- Replace the existing ways to set the fieldManager information
+
+## Proposal
+
+To realize the goals, we would add a field to ObjectMeta called `manager`, right next to the existing `managedFields`.
+This field should not get persisted to storage and is solely for sending fieldManager information to the apiserver.
+When the field is not set, the apiserver would fallback onto current behavior and default the fieldManager (or fail).
+
+This means, we add a field that is optional metadata, write-only and non-persisted.
+The field should take a non empty string or be unset and should follow the same criteria as [ManagedFieldsEntry.Manager](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.14/#managedfieldsentry-v1-meta).
+
+Setting or not setting the field will cause the following behavior for both apply and non-apply operations:
+
+- If the fieldManager option is set for the request (for example through kubectl), it always takes precedence
+- If the newly introduced `manager` field is set in the received request, it will get used as manager
+- If none of the two options match, the fieldManager will default to the user-agent as it currently does
+
+This means, adding the field will only introduce another step for defaulting the fieldManager.
+
+Another option would be, to fail the request if both the field and the fieldManager option are set to different values (like kubectl currently does when setting a different namespace (through `--namespace`) than defined in the manifest file). This might cause more issues with compatibility so the first implementation would stick to the flow outlined above.
```

Comment:
I think we should discuss this, it might even be easier to fail if both values are set but are different.
Don't really have a hard opinion which of the two i favor.